### PR TITLE
Lua generic

### PR DESCRIPTION
**Pull request by bouda1**
I know there are missing tests (in fact they are there but we need an external server for them :-(  ).
I also know there will be several changes needed :-p
So here is a first pull request.

### CODE REVIEW

File: luageneric/inc/com/centreon/broker/luageneric/luabinding.hh [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+                        luabinding(
+                          std::string const& lua_script,
+                          QMap<QString, QVariant> const& conf_params,
+                          macro_cache const& cache);
+                        ~luabinding();
+    bool                has_filter() const;
+    int                 write(misc::shared_ptr<io::data> const& data);
+
+   private:
+
```
Comment (ganoze) [inline luageneric/inc/com/centreon/broker/luageneric/luabinding.hh]: luabinding(luabinding const& other);
luabinding& operator=(luabiding const& other);

File: luageneric/src/luabinding.cc [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+}
+
+/**
+ *  Reads the Lua script, checks its syntax and checks if
+ *   - init()
+ *   - write()
+ *   - filter()
+ *  functions exist in the Lua script. The two first ones are
+ *  mandatory whereas the third one is optional.
+ *
```
Comment (ganoze) [inline luageneric/src/luabinding.cc]: Useless line.

File: luageneric/src/luabinding.cc [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+           << d.type() << " to database query: mapping does not exist");
+}
+
+/**
+ *  Load the Lua interpreter with classical and custom libraries.
+ *
+ * @return The Lua interpreter
+ */
+lua_State* luabinding::_load_interpreter() {
+
```
Comment (ganoze) [inline luageneric/src/luabinding.cc]: Useless \n.

File: luageneric/inc/com/centreon/broker/luageneric/stream.hh [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                    ~stream();
+    stream&         operator=(stream const& other);
+    bool            read(misc::shared_ptr<io::data>& d, time_t deadline);
+    int             write(misc::shared_ptr<io::data> const& d);
+
+  private:
+                    stream(stream const& other);
+
+    // Access to the Lua interpreter
+    std::auto_ptr<luabinding>
```
Comment (ganoze) [inline luageneric/inc/com/centreon/broker/luageneric/stream.hh]: As stream is not copyable, luabinding can be a direct member of the class, without auto_ptr.

File: build/CMakeLists.txt [hunk 5]
```diff
@@ -694,6 +694,21 @@ if (WITH_MODULE_STORAGE)
   list(APPEND MODULE_LIST "storage")
 endif()
 
+# Lua module.
+option(WITH_MODULE_LUAGENERIC "Build luageneric module." ON)
```
[the remark this answers is not in the document]
Comment (bouda1) [reply build/CMakeLists.txt]: Why not...

File: luageneric/src/factory.cc [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+                         bool& is_acceptor,
+                         misc::shared_ptr<persistent_cache> cache) const {
+
+  QMap<QString, QVariant> conf_map;
+  std::string filename(find_param(cfg, "path"));
+  QDomNodeList config = cfg.cfg.elementsByTagName("metrics_lua");
+  for (int i(0); i < config.size(); ++i) {
+    QDomNode conf(config.item(i));
+    QDomNode name = conf.namedItem("name");
+    QDomNode type = conf.namedItem("type");
```
[the remark this answers is not in the document]
Comment (bouda1) [reply luageneric/src/factory.cc]: In fact, at the moment, I have found more people agree with this than with you :-)

File: build/lua/test/socket.lua [hunk 7]
```diff
@@ -0,0 +1,54 @@
+local function check_index(socket)
```
Comment (ganoze) [inline build/lua/test/socket.lua]: Could you move all .lua files that are within the _build_ directory but not related to the build process to the top level _lua_ directory or the top level _tests_ directory ?

File: lua/inc/com/centreon/broker/lua/macro_cache.hh [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+#  include <QHash>
+#  include "com/centreon/broker/instance_broadcast.hh"
+#  include "com/centreon/broker/persistent_cache.hh"
+#  include "com/centreon/broker/neb/instance.hh"
+#  include "com/centreon/broker/neb/host.hh"
+#  include "com/centreon/broker/neb/service.hh"
+#  include "com/centreon/broker/storage/index_mapping.hh"
+#  include "com/centreon/broker/storage/metric_mapping.hh"
+CCB_BEGIN()
```
Comment (ganoze) [inline lua/inc/com/centreon/broker/lua/macro_cache.hh]: Missing \n.

File: lua/src/connector.cc [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+  : io::endpoint(false) {}
+
+/**
+ *  Copy constructor.
+ *
+ *  @param[in] other  Object to copy.
+ */
+connector::connector(connector const& other)
+  : io::endpoint(other),
+    _lua_script(other._lua_script) {}
```
Comment (ganoze) [inline lua/src/connector.cc]: No __conf_params_ or __cache_ copy ? Same stands for assignment operator.

File: lua/src/factory.cc [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+ *
+ *  @return New endpoint.
+ */
+io::endpoint* factory::new_endpoint(
+                         config::endpoint& cfg,
+                         bool& is_acceptor,
+                         misc::shared_ptr<persistent_cache> cache) const {
+  QMap<QString, QVariant> conf_map;
+  std::string filename(find_param(cfg, "path"));
+  QDomNodeList config = cfg.cfg.elementsByTagName("lua_parameters");
```
Comment (ganoze) [inline lua/src/factory.cc]: _lua_parameter_ (with no ending s).

File: storage/test/parser/parse_perfdata.cc [hunk 11]
```diff
@@ -29,6 +30,12 @@ using namespace com::centreon::broker;
 // When parse_perfdata() is called with a valid perfdata string
 // Then perfdata are returned in a list
 TEST(StorageParserParsePerfdata, Simple1) {
+  try {
```
Comment (ganoze) [inline storage/test/parser/parse_perfdata.cc]: Why is this needed ?

Comment (bouda1) [reply storage/test/parser/parse_perfdata.cc]: In case you don't execute all the tests. If you execute all the tests but not Lua ones, the initialization may fail.

File: lua/src/factory.cc (hunk 10 above)
Comment (bouda1) [reply lua/src/factory.cc]: This name was what you proposed in the previous review...
But let's go for 'lua_parameter' now.

Comment (ganoze) [reply lua/src/factory.cc]: Yeah, my bad.

File: storage/test/parser/parse_perfdata.cc (hunk 11 above)
Comment (ganoze) [reply storage/test/parser/parse_perfdata.cc]: Could you give me more details ? This behavior does not seem legit.
